pull --bump: add auto-detection of devel bumps

apjanke opened the pull request:
Not to step on @DomT4's fine work on this, but I had an idea for an alternate way of supporting `brew pull --bump` for `devel` bumps.

Here's a version of `--bump` devel support that auto-detects which of the stable or devel versions, or both, have been changed, and constructs a message accordingly. No need to manually specify the `--devel` flag.

###  Examples  ###

(edited for conciseness)

```
$ brew pull --bump 48465
...
Applying: dummy upgrade acme stable
==> Patch closes issue #48465
==> New bump commit subject: acme 0.93

$ brew pull --bump 48467
...
Applying: dummy upgrade acme devel version
==> Patch closes issue #48467
==> New bump commit subject: acme 0.95 (devel)

$ brew pull --bump 48466
...
Applying: dummy upgrade acme both stable and devel
==> Patch closes issue #48466
==> New bump commit subject: acme 0.96, acme 0.99 (devel)

```

The new helper methods have long-winded names to avoid collisions with other possible methods in `Homebrew`.

Closes #48452

### Testing  ###

Some dummy PRs for testing this stuff against:

#48465 – stable only
#48467 – devel only
#48466 – both stable and devel
#48475 – remove stable
#48488 – bump stable, remove devel
#48478 – new formula "whatever" with only stable 
#49030 – new formula "whatever" with both stable and dev
#48580 – changes formula without bumping stable or devel versions
homebrew/homebrew-versions#1158 – stable only in versions/aspell204


##  Side Note  ##

As far as testing stuff like this goes, I can't wait til core Homebrew code gets split out from the core formula repo. I'm getting my `git branch` on like mad, and often finding myself lost in the forest.

apjanke commented:
Oops; this is currently broken: pulling bottles fails. Fixing...

DomT4 commented:
No worries if you want to step on my toes as you say here :).

My choice to go basic was very deliberate tbh. The pull command is already something of a horror to read & extremely slow on execution and I didn't really want to make either of those problems worse for maintainers or prospective maintainers. Generally I'm fond of simplicity in code heh.

I've only glanced through this because I'm still half asleep but did you handle the situation where there is no stable? It looked like stable was assumed?

apjanke commented:
> I've only glanced through this because I'm still half asleep but did you handle the situation where there is no stable?

Do you mean no change to stable? Or if there was no stable version to start out with?

There was a bug in there wrt no stable version at all. Amended it with a fix just now.

It handles the cases where there stable is unchanged, and if the formula is changed to remove stable (e.g. `acme remove stable, 0.98.76 (devel)`). They look like this, respectively.

```
$ brew pull --bump 48467
...
Applying: dummy upgrade acme devel version
==> New bump commit subject: acme 0.95 (devel)

$ brew pull --bump 48476
Applying: dummy upgrade acme: remove stable, upgrade devel
==> New bump commit subject: acme remove stable, 0.98.76 (devel)
```

mikemcquaid commented:
Makes sense to me. Another case I've been meaning to add for a while is support for `blah 0.1.2 (new formula)` by looking for formula additions. That'd be great if you felt like tackling it in here too?

DomT4 commented:
Alright, closed mine out. I'll be offline for much of today so didn't want to leave it lingering.

> Do you mean no change to stable? Or if there was no stable version to start out with?

I meant the latter; most commonly the `devel-only` tap.

I'll let you and Mike discuss the code side of things, but on a commit style level:

* `==> New bump commit subject: acme 0.96, acme 0.99 (devel)` could do without the duped formula name; that'll take us close to the 50 character recommended commit summary limit in certain cases and it feels redundant.

* `acme remove stable, 0.98.76 (devel)` feels like it should be:
```
acme: remove stable, 0.98.76 (devel)
```
To me, not that I can think of many cases where we've removed stable in the same repo and gone just devel/head-only.

DomT4 commented on Library/Homebrew/cmd/pull.rb:
```diff
@@ -130,12 +217,33 @@ def pull
+          end
+          if old_versions[:devel] != new_versions[:devel]
+            if new_versions[:devel].nil?
+              subject_strs << "remove devel"
```
I'm kind of "eh" about notifying people we've removed devel in favour of stable. It happens a lot and feels obvious that sometimes there won't be a `devel` release available if stable is newer than it.

apjanke commented:
> Another case I've been meaning to add for a while is support for blah 0.1.2 (new formula) by looking for formula additions. That'd be great if you felt like tackling it in here too?

Added support for the new-formula case. If there is a single modified formula, and it was nonexistent/"unavailable" before the patch, it will be marked as a new formula.

```
$ brew pull --bump 48478
...
Applying: dummy new formula "whatever" for testing brew pull --bump
==> New bump commit subject: whatever 1.2.3 (new formula)
```

(I think I can refactor the implementation to collapse most of it in to a single branch; just wanted to get it out there.)

apjanke commented:
> >  Do you mean no change to stable? Or if there was no stable version to start out with?
>
> I meant the latter; most commonly the devel-only tap.

I think this is supported; I just don't have a test case for it. It will compare the old version (nil) to the new version (some string), and since they are different, it will include the "acme X.Y.Z" in the string. The old version is not included in the message, so its exact value doesn't matter.

> ==> New bump commit subject: acme 0.96, acme 0.99 (devel) could do without the duped formula name

Already fixed.

> `    acme remove stable, 0.98.76 (devel)` feels like it should be `acme: remove stable, 0.98.76 (devel)`

Done:

```
$ brew pull --bump 48476
Applying: dummy upgrade acme: remove stable, upgrade devel
==> New bump commit subject: acme: remove stable, 0.98.76 (devel)
```

bfontaine commented:
I’m not sure I understand `get_current_versions_from_info_external`. Why do we need to get version info without loading the formula?

apjanke commented:
> I’m not sure I understand get_current_versions_from_info_external. Why do we need to get version info without loading the formula in the current process?

Because the patch we're about to pull in is going to change the source file for that formula definition, so we need both "before" and "after" info from the same source file. This may just be that I don't know Ruby well enough, but I don't think there's a clean way to "un-require" a file.

apjanke commented:
> I'm kind of "eh" about notifying people we've removed devel in favour of stable. It happens a lot and feels obvious that sometimes there won't be a devel release available if stable is newer than it.

I've no strong feelings either way. Anyone else with preferences here?

bfontaine commented:
Thanks, it makes sense. I’m not sure but there might be a way to reload the file and overwrite the previous formula.

bfontaine commented:
> > I'm kind of "eh" about notifying people we've removed devel in favour of stable. It happens a lot and feels obvious that sometimes there won't be a devel release available if stable is newer than it.
>
> I've no strong feelings either way. Anyone else with preferences here?

I’m on the "eh" side :smiley:

apjanke commented:
> I’m not sure but there might be a way to reload the file and overwrite the previous formula.

I don't think just reloading the file is sufficient, because the `.rb` file might have made arbitrary other changes (defining or extending other functions or classes, changing the state of variables, and so on). We'd need something more extensive to really be clean. (Probably wouldn't matter in most cases, but still...)

apjanke commented:
The ehs have it. Suppressing "remove devel" when stable is also bumped.

```
$ brew pull --bump 48488
Applying: dummy upgrade acme: bump stable, remove devel
==> New bump commit subject: acme 0.94
```

What do y'all think about the `curl` progress bar that displays when downloading the patch? Mind if I get rid of that too, to reduce clutter? Patches should all be small, so a progress bar doesn't seem very useful.

bfontaine commented:
I’ve no strong opinion on this. I do like to have feedback on this but I’ll probably be ok if you remove it.

UniqMartin commented:
Maybe that's beyond the scope of this PR, but what I'd like to see is that `brew pull` checks the subject line of every commit and warns me about commits that don't conform, no matter if I'm using `--bump` or not. I guess is would be pretty straightforward for most changes, as it is either `<formula>: <fix/enhancement>` or `<formula> <new-version>` (possibly suffixed with `(devel)`, `(new formula)`, …). Or is some Git hook a better fit for this kind of check?

apjanke commented:
Seems like that would fit well here, especially since we have the dummy PRs open for testing it. I'll see if I can add it in.

apjanke commented:
Okay, I've added a subject check like Martin requested. Now, for "bumpable" patches – those affecting only a single formula, and no non-formula files, where there was a version change in the formula – it will give you a warning if the (last) commit subject does not match our preferred format.

```
$ brew pull --branch-okay 48466
==> Fetching patch
######################################################################## 100.0%
==> Applying patch
Applying: dummy upgrade acme both stable and devel
Warning: acme has a bottle: do you need to update it with --bottle?
==> Patch closes issue #48466
Warning: Nonstandard bump subject: dummy upgrade acme both stable and devel
Warning: Subject should be: acme 0.96, 0.99 (devel)
==> Patch changed:
 Library/Formula/acme.rb | 10 +++++-----
 1 file changed, 5 insertions(+), 5 deletions(-)
[/usr/local on ⇄ pbump-test-02]
$
```

I added the `--branch-okay` to make testing this easier; I was going a little warning-blind.

Would folks like this check to be the default behavior, or enabled by a command-line option?

apjanke commented:
Okay, I think this is ready to go, pending feedback on the check-subject-when-not-bumping output and any additional testing. I've got the devel behavior in place. Any other comments?

UniqMartin commented on Library/Homebrew/cmd/pull.rb:
```diff
@@ -7,51 +25,16 @@
+    if ARGV.named.empty?
+      odie "This command requires at least one argument containing a URL or pull request number"
+    end
+    do_bump = ((ARGV.include? "--bump") && !(ARGV.include? "--clean"))
```
I find it easier to read if the arguments are in parentheses instead, i.e.:

```ruby
do_bump = ARGV.include?("--bump") && !ARGV.include?("--clean")
```

(But since this is purely a stylistic nit, feel free to ignore this advice.)

apjanke replied:
I agree. Was just having trouble getting the precedence right when I first put this together. Changed.

UniqMartin commented on Library/Homebrew/cmd/pull.rb:
```diff
@@ -88,23 +71,34 @@ def pull
-        "git", "diff-tree", "-r", "--name-only",
-        "--diff-filter=AM", revision, "HEAD", "--", tap.formula_dir.to_s
+          "git", "diff-tree", "-r", "--name-only",
+          "--diff-filter=AM", revision, "HEAD", "--", tap.formula_dir.to_s
```
This indentation change is strange. I don't see what this aligns with.

bfontaine replied:
Seems like a 4-spaces indentation instead of 2-spaces.

apjanke replied:
Yeah. This was a result of me using the "auto-indent" in RubyMine. It defaults to a four-space "continuation indent" inside multiline `(...)` and `{...}`. Fixed.

UniqMartin commented on Library/Homebrew/cmd/pull.rb:
```diff
@@ -88,23 +71,34 @@ def pull
         begin
           changed_formulae << Formula[name]
-        # Make sure we catch syntax errors.
+            # Make sure we catch syntax errors.
```
This indentation change is also strange. (The comment was previously aligned with the `rescue` it most likely refers to.)

apjanke replied:
Another auto-indent discrepancy. Fixed.

UniqMartin commented on Library/Homebrew/cmd/pull.rb:
```diff
@@ -126,42 +120,58 @@ def pull
+        formula = changed_formulae.first
+        new_versions = {
+            stable: formula.stable.nil? ? nil : formula.stable.version.to_s,
+            devel: formula.devel.nil? ? nil : formula.devel.version.to_s
```
I don't think we use this hash syntax anywhere else and it's incompatible with Ruby 1.8 (though that is less of a concern in this particular command). This should be:

```ruby
new_versions = {
  :stable => formula.stable.nil? ? nil : formula.stable.version.to_s,
  :devel => formula.devel.nil? ? nil : formula.devel.version.to_s,
}
```

apjanke replied:
Switched to hash rocket syntax.

UniqMartin commented on Library/Homebrew/cmd/pull.rb:
```diff
@@ -126,42 +120,58 @@ def pull
+        orig_subject = message.empty? ? "" : message.lines.first.chomp
+        subject = subject_for_bump(formula, old_versions, new_versions)
+        if do_bump
+          odie "No version changes found for #{formula.name}" if subject.nil?
```
I think the rest of this `if` branch should be a no-op if `subject == orig_subject`. No need to fix something that's not broken.

apjanke replied:
Agreed.

UniqMartin commented on Library/Homebrew/cmd/pull.rb:
```diff
@@ -126,42 +120,58 @@ def pull
+          odie "No version changes found for #{formula.name}" if subject.nil?
           ohai "New bump commit subject: #{subject}"
-          system "/bin/echo -n #{subject} | pbcopy"
+          system "/bin/echo -n '#{subject}' | pbcopy"
```
Here's a more robust alternative that doesn't rely on a shell, `echo`, and details of shell quoting:

```ruby
Utils.popen_write("pbcopy") { |io| io.write subject }
```

apjanke replied:
Nice!

UniqMartin commented on Library/Homebrew/cmd/pull.rb:
```diff
@@ -126,42 +120,58 @@ def pull
+                              else
+                                "https://github.com/BrewTestBot/homebrew-#{tap.repo}/compare/homebrew:master...pr-#{issue}"
+                              end
+                            end
```
I think this should stay as-is. We prefer this style of aligning with the variable instead of the `if`.

apjanke replied:
Changed. Pulled the `if`on to the next line to allow aligning with the control structure but keeping everything off to the left and close to the variable.

UniqMartin commented on Library/Homebrew/cmd/pull.rb:
```diff
@@ -217,4 +227,141 @@ def pull
+        others << file
+      end
+    end
+    {files: files, formulae: formulae, others: others}
```
Again, use hash rockets to construct a hash.

apjanke replied:
Fixed.

UniqMartin commented on Library/Homebrew/cmd/pull.rb:
```diff
@@ -217,4 +227,141 @@ def pull
+    end
+  end
+
+  def files_changed_in_patch(patchfile)
```
Could this be made a part of `PatchPuller`? (Might require a rename of the class to make sense.)

apjanke replied:
I dunno. To me it makes a bit more sense left outside it, because it could be applied to any patch file, where `PatchPuller` has stuff specific to the lifecycle of a patch file within the `pull` operation (e.g. file location, cleanup). Code-wise, it'd work either way.

UniqMartin commented on Library/Homebrew/cmd/pull.rb:
```diff
@@ -217,4 +227,141 @@ def pull
+      files << $1 if line =~ /^\+\+\+ b\/(.*)/
+    end
+    files.each do |file|
+      if file.start_with?("Library/Formula")
```
I suspect this doesn't work correctly with taps that have their formulae in a different directory?!

apjanke replied:
Crap, you're right. Will need to do a deeper inspection and probably account for edge cases like files being added/moved/deleted. Okay, I'm off to sort that out.

apjanke replied:
Fixed the test to work for taps with the formulae in the root directory.

UniqMartin commented on Library/Homebrew/cmd/pull.rb:
```diff
@@ -217,4 +227,141 @@ def pull
+  end
+
+  def files_changed_in_patch(patchfile)
+    files, formulae, others = [], [], []
```
I think we prefer to split this into multiple lines for readability. (Not entirely sure about that.)

apjanke replied:
[Ruby style guide ](https://github.com/bbatsov/ruby-style-guide#source-code-layout) agrees with you. Changed.

apjanke commented:
Thanks for the feedback! In addition to style stuff, Martin found a bug with it working for taps, so this isn't ready to go yet. I'll sort it out.

In addition, now that this can compare the expected vs actual subject lines, we can avoid sticking redundant bump subject lines on top of commits that are already formatted correctly, and I've enabled that per Martin's suggestion.

UniqMartin commented:
@apjanke Thanks for the quick update! I think I've seen another tap-related issue aside from the one I pointed out, but got distracted and thus stopped reviewing before I reached the end – sorry. And right now it seems better to wait until you to shout *ready* again before doing another review, right?

apjanke commented:
Yeah, give me a bit to work on this so you don't waste your time reviewing an old version.

UniqMartin commented on Library/Homebrew/cmd/pull.rb:
```diff
@@ -217,4 +230,147 @@ def pull
+    versions = {}
+    # Parse version info from "<formula>: stable x.y, ..." `brew info` output
+    # Really this should use the `brew info --json` output, but I don't know Ruby
+    # JSON support yet - apjanke
```
Here's something to experiment with in a `brew irb` session (should get you started):

```ruby
require "utils/json"
json = Utils.popen_read(HOMEBREW_BREW_FILE, "info", "--json=v1", "xz")
puts "unavailable" unless $?.success?
info = Utils::JSON.load(json)
p info[0]["versions"]
```

And if you have `jq` installed, try `brew info --json=v1 xz | jq .` for a pretty-printed output.

apjanke replied:
Thank you! I will give this a try.

apjanke replied:
Switched it to JSON parsing. Thanks!

apjanke commented:
(Latest amendment is just a rebase on master; no significant logic changes.)

apjanke commented:
Okay, I think this is ready to go again.

I fixed the changed-formula-in-patch check to handle taps, and changed the `info` parsing to use JSON. My testing looks good with all of the dummy PRs listed in the summary here.

Any other comments or concerns?

UniqMartin commented on Library/Homebrew/cmd/pull.rb:
```diff
@@ -126,42 +121,61 @@ def pull
+        bottle_commit_url =
+          if testing_job
+            bottle_branch = "testing-bottle-#{testing_job}"
+            url
```
Sorry if I'm being somewhat picky here, but as far as I can tell you haven't changed the code in this and the following two diff chunks, just reformatted it. That's not good for reviewing and for a clean history, so I'd prefer if this was reverted.

UniqMartin commented on Library/Homebrew/cmd/pull.rb:
```diff
@@ -179,9 +193,9 @@ def pull
-              "https://api.bintray.com/content/homebrew/#{repo}/#{package}/#{version}/publish"
+                 "-u#{bintray_user}:#{bintray_key}", "-X", "POST",
+                 "-d", '{"publish_wait_for_secs": -1}',
+                 "https://api.bintray.com/content/homebrew/#{repo}/#{package}/#{version}/publish"
```
Same here. Sorry if I'm being somewhat picky, but you haven't changed the code, just reformatted it. That's not good for reviewing and for a clean history, so I'd prefer if this was reverted.

apjanke replied:
Fair objection. This and the previous one were a result of RubyMine's autoformatting. I thought I had fixed that by changing the "continuation indent" settings, but it didn't take for these. I'll revert.

apjanke replied:
I think I've got this reverted properly in the new amended version. Look good to you now?

UniqMartin replied:
Yes, it's fine now! Thanks! :+1:

UniqMartin commented on Library/Homebrew/cmd/pull.rb:
```diff
@@ -217,4 +231,141 @@ def pull
+      files << $1 if line =~ /^\+\+\+ b\/(.*)/
+    end
+    files.each do |file|
+      if file.start_with?("Library/Formula") || file.match("^[^/]*\.rb")
```
I believe this will still fail for taps like [`homebrew/emacs`](https://github.com/Homebrew/homebrew-emacs) where the formulae reside in the `Formula` directory underneath the repository root. Also, using `start_with?("Library/Formula/")` will avoid the (very unlikely) case that a change is in a file/directory named `Library/Formula<somesuffix>/<somefile>`.

xu-cheng replied:
FWIW, you can use `Tap#formula_dir` abstraction like so:
```
def files_changed_in_patch(tap, patch file)
  ...
  files.each do |file|
    if (tap.path/file).dirname == tap.formula_dir
    ...
  end
  ...
end
```

apjanke replied:
Good points. I didn't know about the `(Homebrew)?Formula` subdir.

I don't think we can use the `Tap#formula_dir` abstraction here, since that looks like it returns the full path to the tap's formula directory as installed on disk, where this method is looking at a patch file with pathnames relative to the repo root. Could do some path munging on that, but if you're okay with doing some simpler path matching, I'd prefer to do it that way, since it lets you operate on just a patch file as input.

xu-cheng replied:
>  since that looks like it returns the full path to the tap's formula directory as installed on disk, where this method is looking at a patch file with pathnames relative to the repo root

That's why I use `tap.path` which returns the root path of the tap.

The reason I prefer to use such abstraction is we support many places to put formula(i.e. root dir, `Formula` and `HomebrewFormula`). And we will drop `Library/Formula` after core formulae separation. So it's better to center all these logic inside `Tap#formula_dir` method. In this way, you won't need to worry the implement detail beneath it.

UniqMartin replied:
I sympathize with @xu-cheng's suggestion, but I'm also fine with the current solution. I don't think the list of supported formula directories for taps will change anytime soon.

xu-cheng replied:
Now think a bit. I found such feature will be used in many places. How about we adding a new method `Tap#formula_file?` as such
```ruby
# tap.rb
class Tap
  def formula_file?(file)
    File.dirname(File.expand_path(file, self.path)) == self.formula_dir.to_s
  end
end

apjanke replied:
Ohhh. xu-cheng's suggestion makes a lot of sense to me now. I think I just didn't read the code correctly the first time. Sorry. Still new to Ruby and getting a hang of how it and the Homebrew extensions deal with files and pathnames.

I'm going to try to rewrite this using `Tap#formula_dir`. My pattern-based solution was a code smell anyway; asking `Tap` makes more sense and is prob more readable. (And because at this point `brew pull`, the target tap is known to be tapped (installed locally), the resolve-to-full-paths approach works (I think).)

`Tap#formula_file?` makes sense to me too, but let's keep it out of this PR so it can avoid touching the higher-risk core library files.

apjanke replied:
@xu-cheng Yep, your original code snippet does exactly what I wanted here, but with better design. Sorry for misreading it first time around. Changing to use it.

UniqMartin commented on Library/Homebrew/cmd/pull.rb:
```diff
@@ -217,4 +231,141 @@ def pull
+        others << file
+      end
+    end
+    {:files => files, :formulae => formulae, :others => others}
```
I believe we prefer to put a space after the opening curly brace `{` and one before the closing curly brace `}`. Have you tried running `brew style` on the changed file?

apjanke replied:
`brew style` agrees. Fixing this and some other nits it picked up (in the code I've added or changed; leaving existing lines alone).

UniqMartin commented on Library/Homebrew/cmd/pull.rb:
```diff
@@ -217,4 +231,141 @@ def pull
+    json = Utils.popen_read(HOMEBREW_BREW_FILE, "info", "--json=v1", formula_name)
+    if $?.success?
+      info = Utils::JSON.load(json)
+      [:stable, :devel, :head].each do |vertype|
```
I believe we can safely skip `:head` here. It's not providing any useful information (the version is always `HEAD`) and we don't deal with `HEAD` version bumps because that wouldn't make any sense. Or am I missing something?

apjanke replied:
I just decided to do all three since those covered all the version cases we support, and make the method more complete. (As an alternative to enumerating the versions present in the JSON and excluding the `bottle` item.) Seems sensible to be completionist when possible and doesn't require much more coding, even if it is a private-use method.

UniqMartin replied:
That's fine with me. Just wanted to point it out. :smile_cat:

UniqMartin commented on Library/Homebrew/cmd/pull.rb:
```diff
@@ -217,4 +231,141 @@ def pull
+      ver_strs = []
+      ver_strs << new[:stable] if new[:stable]
+      ver_strs << "#{new[:devel]} (devel)" if new[:devel]
+      subject = "#{formula.name} #{ver_strs.join(", ")} (new formula)"
```
I think a subject like `awesome-formula 1.1, 1.2alpha1 (devel) (new formula)` looks a bit weird. On the other hand it's probably not worth optimizing for this rather infrequent case, thus feel free to ignore.

apjanke replied:
Yeah, looks a bit weird to me too. Using parentheses for sequential clauses at what are effectively different grouping levels rubs me the wrong way a bit. Bit I think it's still readable, and it's for human consumption, and like you say, infrequent. (And still completely computer-parseable once you know " (new formula)" is a special-case suffix.) No better format immediately springs to mind for me (w/o sacrificing regularity etc).

bfontaine replied:
I think we shouldn’t mention the devel version if there’s a stable one: `awesome-formula 1.1 (new formula)`.

apjanke replied:
Oh yeah, just like if we were bumping the stable version and removing a devel. Changing to do so...

UniqMartin replied:
Yeah, thinking once more about this, I agree with Baptiste.

UniqMartin commented:
Thanks for all your hard work here! The PR is much improved. :clap: 

> My testing looks good with all of the dummy PRs listed in the summary here.

If that's not too much of a hassle, can you add the auto-generated subject lines to this listing?

> Any other comments or concerns?

One major comment and a few nitpicks. Feel free to ignore the latter if you feel I went a little overboard.

apjanke commented:
Okay, I've pushed another amendment, with xu-cheng's improvement to the formula file detection and bfontaine's suggestion for subject lines for new formulae with both stable and dev versions.

Here's a local test run output copy: https://gist.github.com/apjanke/8e3b4c314065b92b0268

Relevant visible changes: new formulae which include both stable and devel versions will look like this, instead of the longer `foo X.Y.Z, XX.YY.ZZ (devel) (new formula)`.
```
==> Patch closes issue #49030
==> New bump commit subject: whatever 1.2.3 (new formula)
```

UniqMartin commented:
:+1: LGTM

apjanke commented:
Merged. Holler if you have any problems.

I'm going to leave those dummy PRs open for a while to help testing any regressions or further PRs on `pull`.

UniqMartin commented:
> Merged.

:clap:
